File: quick_sort.py
import csv

class Data:
    def __init__(self, key, val):
        self.key = key
        self.val = val

def quicksort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        quicksort(arr, low, pi - 1)
        quicksort(arr, pi + 1, high)

def partition(arr, low, high):
    pivot = arr[high].key
    i = low - 1
    for j in range(low, high):
        if arr[j].key <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i+1], arr[high] = arr[high], arr[i+1]
    return i + 1

def read_csv(filename):
    data_list = []
    with open(filename, newline='', encoding='utf-8') as csvfile:
        reader = csv.reader(csvfile)
        for row in reader:
            if len(row) == 2:
                try:
                    key = int(row[0].strip())
                    val = row[1].strip()
                    data_list.append(Data(key, val))
                except ValueError:
                    print(f"Skipping malformed line: {row}")
            else:
                print(f"Skipping malformed line: {row}")
    return data_list

def write_sorted_csv(filename, arr):
    with open(filename, 'w', newline='', encoding='utf-8') as csvfile:
        writer = csv.writer(csvfile)
        for d in arr:
            writer.writerow([d.key, d.val])

def main():
    input_file = 'dataset_sample_1000.csv'
    output_file = 'quick_sort_1000.csv'

    data_list = read_csv(input_file)
    if not data_list:
        print("No data found in the file.")
        return

    quicksort(data_list, 0, len(data_list) - 1)

    write_sorted_csv(output_file, data_list)

    print(f"Sorted data saved to {output_file}")
    print("Sorted array keys:", ' '.join(str(d.key) for d in data_list))

if __name__ == "__main__":
    main()

File: quick_sort.java
import java.io.*;
import java.util.*;

//from data data_sample_10000.csv
public class quick_sort {

    static class Data {
        int key;
        String val;
        Data(int k, String v) { key = k; val = v; }
    }

    public static void quickSort(Data[] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    private static int partition(Data[] arr, int low, int high) {
        int pivot = arr[high].key;
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (arr[j].key <= pivot) {
                i++;
                swap(arr, i, j);
            }
        }
        swap(arr, i + 1, high);
        return i + 1;
    }

    private static void swap(Data[] arr, int i, int j) {
        if (i != j) {
            Data temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }

    private static List<Data> readCSV(String filename) throws IOException {
        List<Data> list = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",", 2);
                if (parts.length == 2) {
                    try {
                        int key = Integer.parseInt(parts[0].trim());
                        String val = parts[1].trim();
                        list.add(new Data(key, val));
                    } catch (NumberFormatException e) {
                        System.err.println("Skipping malformed integer line: " + line);
                    }
                } else {
                    System.err.println("Skipping malformed line: " + line);
                }
            }
        }
        return list;
    }

    private static void writeSortedCSV(String filename, Data[] arr) throws IOException {
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(filename))) {
            for (Data d : arr) {
                bw.write(d.key + "," + d.val);
                bw.newLine();
            }
        }
    }

    public static void main(String[] args) {
        String csvFile = "dataset_sample_1000.csv"; // fixed input filename
        String outputCsvFile = "quick_sort_1000.csv";

        try {
            List<Data> dataList = readCSV(csvFile);
            if (dataList.isEmpty()) {
                System.err.println("No data found in the file.");
                return;
            }
            Data[] arr = dataList.toArray(new Data[0]);

            quickSort(arr, 0, arr.length - 1);

            writeSortedCSV(outputCsvFile, arr);
            System.out.println("Sorted data saved to " + outputCsvFile);

            System.out.print("Sorted array keys: ");
            for (Data d : arr) System.out.print(d.key + " ");
            System.out.println();

        } catch (IOException e) {
            System.err.println("IO Error: " + e.getMessage());
        }
    }
}
